File: branches/darc/Java/OPSJLib/src/ops/Notifier.java
/*
 * To change this template, choose Tools | Templates
 * and open the template in the editor.
 */

package ops;

import java.util.Vector;

/**
 *
 * @author Anton Gravestam
 */
public class Notifier <T>
{
    Vector<Listener<T>> listeners = new Vector<Listener<T>>();

    void notifyListeners(T arg)
    {
        for (Listener<T> listener : listeners)
        {
            listener.onNewEvent(this, arg);
        }
    }

    public boolean removeListener(Listener<T> listener)
    {
        return listeners.remove(listener);
    }

    public synchronized boolean addListener(Listener<T> listener)
    {
        return listeners.add(listener);
    }


}

File: C#/Ops/Notifier.cs
///////////////////////////////////////////////////////////
//  Notifier.cs
//  Implementation of the Class Notifier
//  Created on:      12-nov-2011 09:25:32
//  Author:
///////////////////////////////////////////////////////////

using System.Collections.Generic;
using System.Runtime.CompilerServices;  // Needed for the "MethodImpl" synchronization attribute

namespace Ops 
{
	public class Notifier<T> 
    {
		internal List<IListener<T>> listeners = new List<IListener<T>>();

        public void NotifyListeners(T arg)
        {
            foreach (IListener<T> listener in listeners)
            {
                listener.OnNewEvent(this, arg);
            }
        }

        public void RemoveListener(IListener<T> listener)
        {
            listeners.Remove(listener);
        }

        [MethodImpl(MethodImplOptions.Synchronized)]
        public void AddListener(IListener<T> listener)
        {
            listeners.Add(listener);
        }

	}

}